widget/tabcontainer_test.go

```diff
@@ -47,12 +51,77 @@ func TestTabContainer_SelectTabIndex(t *testing.T) {
 	assert.Equal(t, 1, tabs.CurrentTabIndex())
 }
 
-func TestTabContainer_ApplyTheme(t *testing.T) {
-	tabs := NewTabContainer(&TabItem{Text: "Test1", Content: NewLabel("Test1")})
-	underline := Renderer(tabs).(*tabContainerRenderer).line
-	barColor := underline.FillColor
+func TestTabContainerRenderer_ApplyTheme(t *testing.T) {
+	tabs := widget.NewTabContainer(widget.NewTabItem("Test1", widget.NewLabel("Test1")))
+	var underline *canvas.Rectangle
+	driver.WalkObjectTree(tabs, fyne.NewPos(0, 0), func(o fyne.CanvasObject, offset fyne.Position) bool {
```

Comment:
I don't quite follow why we are going to all the effort of traversing a widget tree to avoid accessing the tabContainerRenderer fields? I see that this has moved to a _test package - but as renderers are internal details this seems like extra work.
I agree with the _test package for main widget tests where the test are for the main API but here we are testing internal details...

Reply:
I don't think so. That the color of the line changes on theme switch is an expectation to the widget's representation.
To be honest, I think that we should even check, that the background color of the tabs, the text color and the icon color change, too.
I would put the test more in this direction:
```go
func TestTabContainer_ChangingTheme(t *testing.T) {
…
  	fyne.CurrentApp().Settings().SetTheme(theme.DarkTheme())
	assert.Equal(t, theme.DarkTheme().BackgroundColor(), underline.FillColor)

	fyne.CurrentApp().Settings().SetTheme(theme.LightTheme())
	assert.Equal(t, theme.LightTheme().BackgroundColor(), underline.FillColor)
}
```

The call to `ApplyTheme()` is not needed.

Reply:
It may be an expectation of the representation but it is not part of the widget's API or contract.
The reason that we use ApplyTheme is because there is no guarantee that setting a theme will immediately change all widgets - that is handled using a settings change listener and applies across the widget tree in order. As such the assertion could fail due to the speed of the code executing.

I guess my overall question is - why is such a test style change applied to the entire file as part of a PR to add functionality to tabs? If we are making stylistic changes we should keep them separate from the functional/API changes.

Reply:
As I already said: To use the object tree walker it was necessary to move to the test package. Thus I had no longer access to the tab container renderer nor its line.

However, I reverted this and changed the layout test to use the knowledge about the renderer structure. It'll probably bite me when I add support for vertical tab bar but so it might be.

Reply:
I appreciate that this is a compromise - let's review if/when it is required for the feature addition.